### PR TITLE
path argument is used to create an intersection with existing finder

### PR DESCRIPTION
**Pull request by keradus**
ref #1719 

@tifabien , please, could you heavily play with this change?
(especially from CLI user side and your hooks)

### CODE REVIEW

**Comment by tifabien** [comment]
I'll test it asap but at the first look it seems it will solve my problem :) 
Thanks a lot!

**Comment by keradus** [comment]
General question. 
We could backport this change to 1.x as well (ref https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/1.12/Symfony/CS/Console/Command/FixCommand.php#L356)
It may be considered as behaviour change, and for that BC break - because one asked for fixing excluded file and it won't be fixed.
Also, it may be considered as bug fix, so not BC break - because one asked to fix file, use configuration that exclude this file so finally it should not be changed.

Any thoughts about change itself and then target version ?
cc @stof @sstok @GrahamCampbell @SpacePossum @gharlan @ceeram and everybody  ;)

**Comment by GrahamCampbell** [comment]
I think it's fine on 2.x only.

**Comment by keradus** [comment]
It shouldn't. Please measure and show.
We will need to do it that way on 2.x when we will have cachingIterator to lint files before fixing them, and we need to ommit some files before linting

**Comment by tifabien** [comment]
:+1: I tested my side and it works like a charm :) 

Small question, Do you know when you'll plan to release the 2.x?

Thanks a lot !

**Comment by keradus** [comment]
Great to hear!
I'm still waiting for some more opinions on this PR.

About release - I won't give you exact date, but I think we are close:
https://github.com/FriendsOfPHP/PHP-CS-Fixer/milestones/v2.0

File: src/FileCacheManager.php [hunk 1]
```diff
@@ -49,6 +50,7 @@ public function __construct($isEnabled, $cacheFile, array $rules)
     {
         $this->isEnabled = $isEnabled;
         $this->cacheFile = $cacheFile;
+        $this->cacheFileRealDirName = dirname(realpath($cacheFile));
```
Comment (SpacePossum) [inline src/FileCacheManager.php]: does `realpath` work (correctly) in `phar`'s? (I recall there were some issues with it)

File: tests/Console/ConfigurationResolverTest.php [hunk 2]
```diff
@@ -246,11 +246,53 @@ public function testResolveConfigFileChooseFileWithInvalidFile()
...
+
+        $this->assertCount(0, iterator_to_array($this->resolver->getConfig()->getFinder()));
+    }
+
+    public function testResolvePathWithFileThatIsExcludedByDir()
+    {
+        $dir = realpath(__DIR__.'/..');
+        $this->config->getFinder()
+            ->in($dir)
+            ->exclude(substr(realpath(__DIR__), strlen($dir) + 1));
```
Comment (gharlan) [inline tests/Console/ConfigurationResolverTest.php]: is realpath necessary here?

File: tests/Console/ConfigurationResolverTest.php [hunk 3]
```diff
@@ -246,11 +246,53 @@ public function testResolveConfigFileChooseFileWithInvalidFile()
...
+        $dir = realpath(__DIR__);
+        $this->config->getFinder()
+            ->in($dir)
+            ->notPath(substr(__FILE__, strlen($dir) + 1));
+
+        $this->resolver
+            ->setOption('path', __FILE__)
+            ->resolve();
+
+        $this->assertCount(0, iterator_to_array($this->resolver->getConfig()->getFinder()));
```
Comment (gharlan) [inline tests/Console/ConfigurationResolverTest.php]: no need to call `iterator_to_array`, `assertCount` can handle traversables.

Comment (SpacePossum) [inline tests/Console/ConfigurationResolverTest.php]: but I doesn't handle iterators well on all versions (https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1632#issuecomment-167264865), we could try this case out though :)

Comment (gharlan) [inline tests/Console/ConfigurationResolverTest.php]: interesting. didn't know that.

**Comment by gharlan** [comment]
> Any thoughts about change itself and then target version ?

:+1: for change itself, but only 2.x

**Comment by SpacePossum** [comment]
Besides https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1742#discussion_r52050731 I would like to have a little explaining about the new behavior if that is ok :)
(maybe we could at it to the docs somewhere as well?)

If I understand the issue correctly together with the PR; the fixer should only fix the given file / files in the directory passed on by command line argument, but still honor the finder exclude rules set in the config file (and fixers/level)?
 (and not apply possibly broader finder rules that would be out of the directory passed on by the command line?)

**Comment by keradus** [comment]
@SpacePossum 
it would create an intersection 
https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/1742/files#diff-62b374855f5961454b817b144c531e27R425

So, if you run the tool with both finder in configuration and path as CLI arguments only the files that would be found by both finders will be fixed.

**Comment by SpacePossum** [comment]
Thanks for the explaining :)
So if one would black list directory `a/b/c` in its config and than run the fixer using `a/b/c` as input path by the command line, what would happen?

**Comment by keradus** [comment]
Intersection would be empty.
That's the point of @tifabien #1719

If one configure his project to not touch sth, it should not be touched. If one decided to change that configuration - that's what config file is for.
Also remember that one can't exclude path by CLI.

File: tests/Console/ConfigurationResolverTest.php (hunk 2 above)
Comment (keradus) [inline tests/Console/ConfigurationResolverTest.php]: it will normalize that `/..` part

**Comment by SpacePossum** [comment]
I see, I would like the command line input to be leading in this case. I thought the issue reported was something like;
- config blacklists dir `a/b/c`
- run with command line `a/b`
- `a/b/c` is still being fixed

your PR will fix that, so :+1:  :)
However as a result of the PR, it will be impossible to fix files in `a/b/c` using the command line without changing the config file. If one would explicitly pass `a/b/c` as command line arg, I would expected the fixer to ran over dir. Its a tricky one, I think both solutions have + and -, so I'm still :+1: for the change as it fixes the issue for some people and +0 for the side effect :)

File: src/FileCacheManager.php (hunk 1 above)
Comment (keradus) [inline src/FileCacheManager.php]: I never had any problems with it, could you elaborate?

Comment (SpacePossum) [inline src/FileCacheManager.php]: Cant recall any details, sorry man, ignore my comment

File: tests/Console/ConfigurationResolverTest.php (hunk 2 above)
Comment (gharlan) [inline tests/Console/ConfigurationResolverTest.php]: I can not see that part here.. ;)

Comment (keradus) [inline tests/Console/ConfigurationResolverTest.php]: if `__DIR__would be in linked path then realpath will reveal real path that link point to.

If you have project forked in non-linked path then indeed, `$dir = realpath(__DIR__.'/..');` will only move one dir up and then second realpath is not needed. But if you would have in in linked path, then first realpath will reveal linked point, so the second one is needed.

If you have some nice way of normalizing `/foo/bar/.././baz` into `/foo/baz` I'm willing to listen, as it's the only reason I called realpath here.

File: src/FileCacheManager.php (hunk 1 above)
Comment (keradus) [inline src/FileCacheManager.php]: OK, I think I may know what was in your mind.
calling realpath on phar:// stream won't provide useful data, but calling it (even inside phar file) on file that is outside phar (like here) is ok

**Comment by keradus** [comment]
Done from my side.
It would be really nice to close it fast ;)

File: tests/Console/ConfigurationResolverTest.php (hunk 2 above)
Comment (gharlan) [inline tests/Console/ConfigurationResolverTest.php]: hmm, symlinks should be resolved in `__DIR__`, so it should be always equivalent to `realpath(__DIR__)`.

http://php.net/manual/en/language.constants.predefined.php

`__FILE__`:
> The full path and filename of the file **with symlinks resolved**.

`__DIR__`:
> This is equivalent to `dirname(__FILE__)`.

File: .travis.yml [hunk 4]
```diff
@@ -22,7 +22,8 @@ cache:
 
 before_install:
     - git config --global github.accesstoken 5e7538aa415005c606ea68de2bbbade0409b4b8c
-    - 'if [ "$SYMFONY_VERSION" != "" ]; then sed -i "s/\"symfony\/\([^\"]*\)\": \"[^\"]*\"/\"symfony\/\1\": \"$SYMFONY_VERSION\"/g" composer.json; fi'
+    - 'if [ "$SYMFONY_VERSION" != "" ]; then sed -i "s/\"symfony\/\([^\"]*\)\": \"^2[^\"]*\"/\"symfony\/\1\": \"$SYMFONY_VERSION\"/g" composer.json; fi'
+    - cat composer.json
```
Comment (gharlan) [inline .travis.yml]: is this change intended in this pr?

Comment (gharlan) [inline .travis.yml]: ah, it is needed for the polyfill component, right?

Comment (keradus) [inline .travis.yml]: exactly

File: tests/Console/ConfigurationResolverTest.php (hunk 2 above)
Comment (keradus) [inline tests/Console/ConfigurationResolverTest.php]: thanks!

**Comment by gharlan** [comment]
@keradus please take a look at https://github.com/keradus/PHP-CS-Fixer/pull/15

File: tests/Console/ConfigurationResolverTest.php [hunk 5]
```diff
@@ -245,11 +244,53 @@ public function testResolveConfigFileChooseFileWithInvalidFile()
...
+
+        $this->assertSame(__DIR__, $this->resolver->getPath());
+    }
+
+    public function testResolvePathWithFileThatIsExcludedDirectly()
+    {
+        $dir = __DIR__;
+        $this->config->getFinder()
+            ->in($dir)
+            ->notPath(substr(__FILE__, strlen($dir) + 1));
```
Comment (gharlan) [inline tests/Console/ConfigurationResolverTest.php]: isn't it the same as `basename(__FILE__)`?

File: UPGRADE.md [hunk 6]
```diff
@@ -34,6 +34,12 @@ CLI options
               | --rules       | Rules to be used            | option was added
               | --using-cache | Does cache should be used   | option was added
 
+CLI argument
+------------
+
+On 1.x line `path` argument allow you to specify the dir/file that will be fixed.
```
Comment (gharlan) [inline UPGRADE.md]: "allows"

File: tests/Console/ConfigurationResolverTest.php [hunk 7]
```diff
@@ -245,11 +244,52 @@ public function testResolveConfigFileChooseFileWithInvalidFile()
...
+
+        $this->assertCount(0, $this->resolver->getConfig()->getFinder());
+    }
+
+    public function testResolvePathWithFileThatIsExcludedByDir()
+    {
+        $dir = realpath(__DIR__.'/..');
+        $this->config->getFinder()
+            ->in($dir)
+            ->exclude(substr(__DIR__, strlen($dir) + 1));
```
Comment (gharlan) [inline tests/Console/ConfigurationResolverTest.php]: `basename(__DIR__)`

File: tests/Console/ConfigurationResolverTest.php [hunk 8]
```diff
@@ -245,11 +244,52 @@ public function testResolveConfigFileChooseFileWithInvalidFile()
...
+
+        $this->assertCount(0, $this->resolver->getConfig()->getFinder());
+    }
+
+    public function testResolvePathWithFileThatIsNotExcluded()
+    {
+        $dir = __DIR__;
+        $this->config->getFinder()
+            ->in($dir)
+            ->notPath('foo-'.substr(__FILE__, strlen($dir) + 1));
```
Comment (gharlan) [inline tests/Console/ConfigurationResolverTest.php]: basename

File: tests/Console/ConfigurationResolverTest.php [hunk 9]
```diff
@@ -245,11 +244,52 @@ public function testResolveConfigFileChooseFileWithInvalidFile()
...
+        $this->resolver
+            ->setOption('path', __FILE__)
+            ->resolve();
+
+        $this->assertCount(0, $this->resolver->getConfig()->getFinder());
+    }
+
+    public function testResolvePathWithFileThatIsExcludedByDir()
+    {
+        $dir = realpath(__DIR__.'/..');
```
Comment (gharlan) [inline tests/Console/ConfigurationResolverTest.php]: `dirname(__DIR__)`

**Comment by keradus** [comment]
> @keradus please take a look at keradus#15

merged, thank you @gharlan !

**Comment by SpacePossum** [comment]
looking good :)
:+1:

File: src/Console/ConfigurationResolver.php [hunk 10]
```diff
@@ -390,13 +392,51 @@ private function resolveConfig()
...
+                $callback
+            );
+        } catch (\LogicException $e) {
+            $iterator = $isFile
+                ? new \ArrayIterator(array(new SymfonySplFileInfo(
+                    $this->path,
+                    substr(dirname($this->path), strlen($this->cwd) + 1),
+                    substr($this->path, strlen($this->cwd) + 1)
+                )))
+                : $currentFinder->in($this->path)
```
Comment (gharlan) [inline src/Console/ConfigurationResolver.php]: Isn't it possible that `$currentFinder` is not a symfony finder instance, so the method `in` would not exist?

**Comment by keradus** [comment]
looks ready

**Comment by gharlan** [comment]
:+1:

**Comment by keradus** [comment]
Thank you @gharlan.

**Comment by gharlan** [comment]
thanks!
